build/configure.js

```diff
@@ -0,0 +1,156 @@
+// AppConfigureStep is per app configure step generator.
+var AppConfigureStep = require('./configure/app_configure_step');
+
+// Default building steps before app's own building steps.
+const PRE_APP_BUILD_STEPS = [
+  'svoperapps',
+  'webapp-manifests',
+  'contacts-import-services',
+  'search-provider',
+  'keyboard-layouts',
+  'preferences',
+  'settings',
+  'webapp-shared',
+  'copy-common-files'
+];
+
+/**
+ * ConfigureStep would generate all the data for Buildbackend. Currently would
+ * contain PRE_APP_BUILD_STEPS and each app's steps.
+ * @param {object} options - build_config from main Makefile.
+ * @constructor
+ */
+var ConfigureStep = function(options) {
+  this.options = options;
+};
+
+ConfigureStep.prototype = {
+  /**
+   * Configure start from here.
+   */
+  start: function() {
+    var commonMkPath = utils.joinPath(this.options.GAIA_DIR, 'all.mk');
+    this.mainMake = new DependencyGraph(commonMkPath);
+
+    // TODO: we should be able to detect whether to regenerate makefile.
+    this.preAppConfig();
+    this.postAppConfig();
+    // build_config.in has contained all necessary ENV data for the generated
+    // makefile.
+    this.mainMake.genBackend('-include build_config.in\n');
+
+    // Execute the generated makefile directly.
+    this.mainMake.executeBackend();
+  },
+
+  /**
+   * Generate config data from PRE_APP_BUILD_STEPS.
+   */
+  preAppConfig: function() {
+    this._stepIndex = this.options.STAGE_DIR;
+    PRE_APP_BUILD_STEPS.forEach(function(step) {
+      // We don't generate setting.js if only build one app.
+      if (this.options.BUILD_APP_NAME !== '*' && step === 'settings') {
+        return;
+      }
+      this.mainMake.insertTask(
+        'phony',
+        'build_' + step,
+        [this._stepIndex],
+        '@$(call $(BUILD_RUNNER),' + step + ')'
+      );
+      this._stepIndex = 'build_' + step;
+    }, this);
+  },
+
+  /**
+   * Check whether the app should be reconfig or not.
+   * @param {string} appDir - the app folder path.
+   */
+  shouldBuild: function(appDir) {
+    var buildAppName = this.options.BUILD_APP_NAME;
+    if (buildAppName === '*') {
+      return true;
+    } else if (buildAppName === 'callscreen' &&
+               appDir.indexOf('communications') !== -1) {
+    // A workaround for bug 1093267 in order to handle callscreen's l10n
+    // broken. Callscreen will generate incorrect multilocale strings if
+    // build_stage/communications/dialer/locales is removed by
+    // webapp-optimize. After bug 1093267 has been resolved, we're going to
+    // get rid of this.
+      return true;
+    } else {
+      return utils.getAppNameRegex(this.options.BUILD_APP_NAME).test(appDir);
+    }
+  },
+  /**
+   * Generate config data from app's AppConfigureStep and set it depends on
+   * build_profile.
+   */
+  postAppConfig: function() {
+    this.mainMake.insertTask(
+      'phony',
+      'pre_stage',
+      [this._stepIndex]
+    );
+
+    this.mainMake.insertTask(
```

Comment:
Sorry, correct myself. 
I still need this line to put profile to the top of each makefile, so that when we execute make -f app.mk, it will go all the steps.

ps: I've also checked .DEFAULT and all, they all not work.